File: UTPR_Matlab_Model/utpr_0520/utpr_lpv.cpp
#include <vector>
#include <cmath>

// utpr_lpv 函数
void utpr_lpv(const std::vector<double>& x, double T, std::vector<std::vector<double>>& A, std::vector<std::vector<double>>& B) {
    // 定义物理参数
    double l_1 = 0.34;
    double lg_1 = 0.17;
    double l_2 = 0.29;
    double lg_2 = 0.145;
    double l_3 = 0.52;
    double lg_3 = 0.26;
    double m1 = 1.258;
    double m2 = 5.686;
    double m3 = 2.162;
    double J1 = 0.0121;
    double J2 = 0.0398;
    double J3 = 0.0487;
    double g = 9.81;

    // 从状态变量中提取角度和角速度
    std::vector<double> theta = {x[0], x[1], x[2]};
    std::vector<double> omega = {x[3], x[4], x[5]};

    // 计算中间变量
    double a1 = J1 + m1 * lg_1 * lg_1 + (m2 + m3) * l_1 * l_1;
    double a2 = J2 + m2 * lg_2 * lg_2 + m3 * l_2 * l_2;
    double a3 = (m2 * lg_2 + m3 * l_2) * l_1;
    double a4 = J3 + m3 * lg_3 * lg_3;
    double a5 = m3 * l_1 * lg_3;
    double a6 = m3 * l_2 * lg_3;

    double b1 = (m1 * lg_1 + m2 * l_1 + m3 * l_1) * g;
    double b2 = (m2 * lg_2 + m3 * l_2) * g;
    double b3 = (m3 * lg_3) * g;

    // 初始化矩阵 A 和 B
    A.resize(3, std::vector<double>(3));
    B.resize(3, std::vector<double>(3));

    // 计算矩阵 A 和 B
    A[0][0] = a1 + a2 + a4 + 2 * a5 * cos(theta[1] + theta[2]) + 2 * a3 * cos(theta[1]) + 2 * a6 * cos(theta[2]);
    A[0][1] = a2 + a4 + a3 * cos(theta[1]) + a5 * cos(theta[1] + theta[2]) + 2 * a6 * cos(theta[2]);
    A[0][2] = a4 + a5 * cos(theta[1] + theta[2]) + a6 * cos(theta[2]);
    A[1][0] = A[0][1];
    A[1][1] = a2 + a4 + 2 * a6 * cos(theta[2]);
    A[1][2] = a4 + a6 * cos(theta[2]);
    A[2][0] = A[0][2];
    A[2][1] = A[1][2];
    A[2][2] = a4;

    B[0][0] = -b1 * sin(theta[0]) - b2 * sin(theta[0] + theta[1]) - b3 * sin(theta[0] + theta[1] + theta[2]);
    B[1][0] = -b2 * sin(theta[0] + theta[1]) - b3 * sin(theta[0] + theta[1] + theta[2]);
    B[2][0] = -b3 * sin(theta[0] + theta[1] + theta[2]);
}

File: utpr_lpv.m
function [A, B, M] = utpr_lpv(x, T)

g = 9.81; % Force of gravity
l_1 = 0.325;  % Length of the arm
lg_1 = 0.245436; % Center of gravity of the arm
l_2 = 0.2; % Length of the pendulum
lg_2 = 0.165468;
l_3 = 0.2; % Length of the pendulum
lg_3 = 0.127337;
m1 = 1.915; % Mass of the arm
m2 = 1.469; % Mass of the pendulum
m3 = 1.141; % Mass of the pendulum
J1 = 0.025720752;
J2 = 0.007667511;
J3 = 0.004949014;
% g = 9.81;


%% Disect states
% For easier calculation in the following lines, we split the states into theta and its derivative
theta = x(1:3);
omega = x(4:6);

%% intermediate variable
a1 = J1+m1*lg_1^2+(m2+m3)*l_1^2;
a2 = J2+m2*lg_2^2+m3*l_2^2;
a3 = (m2*lg_2+m3*l_2)*l_1;
a4 = J3+m3*lg_3^2;
a5 = m3*l_1*lg_3;
a6 = m3*l_2*lg_3;

b1 = (m1*lg_1+m2*l_1+m3*l_1)*g;
b2 = (m2*lg_2+m3*l_2)*g;
b3 = (m3*lg_3)*g;
%% Build preliminary matrices
% These are the matrices used to define also the nonlinear model in the paper. The vector G is not
% defined here, as it is replaced by the matrix K
M  = [ a1+a2+a4+2*a5*cos(theta(2)+theta(3))+2*a3*cos(theta(2))+2*a6*cos(theta(3)), ... 
        a2+a4+a3*cos(theta(2))+a5*cos(theta(2)+theta(3))+2*a6*cos(theta(3)), ...
        a4+a5*cos(theta(2)+theta(3))+a6*cos(theta(3));
        a2+a4+a3*cos(theta(2))+a5*cos(theta(2)+theta(3))+2*a6*cos(theta(3)), ...
        a2+a4+2*a6*cos(theta(3)), ...
        a4+a6*cos(theta(3));
        a4+a5*cos(theta(2)+theta(3))+a6*cos(theta(3)), ...
        a4+a6*cos(theta(3)), ...
        a4  ];

C  = [  -a5*(omega(2)+omega(3))*sin(theta(2)+theta(3))-a3*omega(2)*sin(theta(2))-a6*omega(3)*sin(theta(3)), ...
        -a5*(omega(1)+omega(2)+omega(3))*sin(theta(2)+theta(3))-a3*(omega(1)+omega(2))*sin(theta(2))-a6*omega(3)*sin(theta(3)), ...
        -(omega(1)+omega(2)+omega(3))*(a5*sin(theta(2)+theta(3))+a6*sin(theta(3)));
        a5*omega(1)*sin(theta(2)+theta(3))+a3*omega(1)*sin(theta(2))-a6*omega(3)*sin(theta(3)), ...
        -a6*omega(3)*sin(theta(3)), ...
        -a6*(omega(1)+omega(2)+omega(3))*sin(theta(3));
        a5*omega(1)*sin(theta(2)+theta(3))+a6*(omega(1)+omega(2))*sin(theta(3)), ...
        a6*(omega(1)+omega(2))*sin(theta(3)), ...
        0  ];
% K matrix
% K = [-b1*si(theta(1))-b2*si(theta(1)+theta(2))-b3*si(theta(1)+theta(2)+theta(3)), ...
%      -b2*si(theta(1)+theta(2))-b3*si(theta(1)+theta(2)+theta(3)), ...
%      -b3*si(theta(1)+theta(2)+theta(3));
%      -b2*si(theta(1)+theta(2))-b3*si(theta(1)+theta(2)+theta(3)), ...
%      -b2*si(theta(1)+theta(2))-b3*si(theta(1)+theta(2)+theta(3)), ...
%      -b3*si(theta(1)+theta(2)+theta(3));
%      -b3*si(theta(1)+theta(2)+theta(3)), ...
%      -b3*si(theta(1)+theta(2)+theta(3)), ...
%      -b3*si(theta(1)+theta(2)+theta(3)) ];

K = [-b1-b2-b3, ...
     -b2-b3, ...
     -b3;
     -b2-b3, ...
     -b2-b3, ...
     -b3;
     -b3, ...
     -b3, ...
     -b3 ];


% K = [-b1*si(theta(1))-b2*sin(theta(1)+theta(2))*(1/theta(1))-b3*sin(theta(1)+theta(2)+theta(3))*(1/theta(1)),   0,     0;...
%      0,   -b2*sin(theta(1)+theta(2))*(1/theta(2))-b3*sin(theta(1)+theta(2)+theta(3))*(1/theta(2)),      0;...
%      0,   0,      -b3*sin(theta(1)+theta(2)+theta(3))*(1/theta(3)) ];



% We cannot avoid inverting M, as it is needed for the input matrix B
Mi  = inv(M);
if any(~isfinite(Mi(:)))
    error('Hessian matrix Mi contains non-finite values.');
end

%% Assemble model matrices from parts
% From the previously defined parts, assemle the final model matrices A & B. At the same time, apply
% Euler's method for descritization.

A  = [  zeros(3)   eye(3) ;
       -Mi*K      -Mi*C   ];
B  = [ zeros(3,1)  zeros(3,1);
       Mi(:, 2)   Mi(:, 3)];
 
% Discretize if sampling time is given
if nargin >= 2
    A   = eye(size(A)) + T*A;
    B   = T*B;
end

if any(~isfinite(A(:)))
    error('Hessian matrix A contains non-finite values.');
end

if any(~isfinite(B(:)))
    error('Hessian matrix B contains non-finite values.');
end

if any(~isfinite(Mi(:)))
    error('Hessian matrix Mi contains non-finite values.');
end

end

function y = si(x)
%SI Defines the SI function
%   Matlab does not define the SI function, so we need to. Just taking sin(x)/x will result in a
%   singularity for x = 0, which is undesirable

    if x == 0
        y = 1;
    else
        y = sin(x) / x;
    end
end
